MSE for Audio

samdutton opened the pull request:
Ported from Dale Curtis's article [here](http://dalecurtis.github.io/llama-demo/index.html).

Three examples are includes in iframes from [simpl.info/mse/audio/gap](https://simpl.info/mse/audio/gap), etc.

This article should eventually be moved to /web/audio.

PaulKinlan commented:
@petele can we include iframes in the main site?

On Fri, 19 Jun 2015 17:52 Sam Dutton <notifications@github.com> wrote:

> Ported from Dale Curtis's article here
> <http://dalecurtis.github.io/llama-demo/index.html>.
>
> Three examples are includes in iframes from simpl.info/mse/audio/gap, etc.
>
> This article should eventually be moved to /web/audio.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/google/WebFundamentals/pull/1381
> Commit Summary
>
>    - Added MSE for Audio post
>
> File Changes
>
>    - *A*
>    src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown
>    <https://github.com/google/WebFundamentals/pull/1381/files#diff-0>
>    (373)
>    - *A*
>    src/_langs/en/updates/images/2015-06-12-media-source-extensions-for-audio/mp3_gap.png
>    <https://github.com/google/WebFundamentals/pull/1381/files#diff-1> (0)
>    - *A*
>    src/_langs/en/updates/images/2015-06-12-media-source-extensions-for-audio/mp3_gap_end.png
>    <https://github.com/google/WebFundamentals/pull/1381/files#diff-2> (0)
>    - *A*
>    src/_langs/en/updates/images/2015-06-12-media-source-extensions-for-audio/mp3_mid.png
>    <https://github.com/google/WebFundamentals/pull/1381/files#diff-3> (0)
>
> Patch Links:
>
>    - https://github.com/google/WebFundamentals/pull/1381.patch
>    - https://github.com/google/WebFundamentals/pull/1381.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/WebFundamentals/pull/1381>.
>

jpmedley commented on src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown:
```diff
@@ -0,0 +1,373 @@
+{% endhighlight %}
+
+Once the MediaSource object is connected, it will perform some initialization and eventually fire a "sourceopen" event; at which point we can create a [SourceBuffer](http://www.w3.org/TR/media-source/#sourcebuffer). In the example above, we're creating an "audio/mpeg" one, which is able to parse and decode our MP3 segments; there are several [other types](http://www.w3.org/2013/12/byte-stream-for ...
+
```
Please replace quotes on sourceopen with backticks(`) to give it a different type face.

samdutton replied:
Good spot! Found a few of these. Changed.

On 19 June 2015 at 18:24, Joe Medley <notifications@github.com> wrote:

> In
> src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown
> <https://github.com/google/WebFundamentals/pull/1381#discussion_r32849947>
> :
>
> > +  function onAudioLoaded(data, index) {
> > +    // Append the ArrayBuffer data into our new SourceBuffer.
> > +    sourceBuffer.appendBuffer(data);
> > +  }
> > +
> > +  // Retrieve an audio segment via XHR.  For simplicity, we're retrieving the
> > +  // entire segment at once, but we could also retrieve it in chunks and append
> > +  // each chunk separately.  MSE will take care of assembling the pieces.
> > +  GET('sintel/sintel_0.mp3', function(data) { onAudioLoaded(data, 0); } );
> > +}, false);
> > +
> > +audio.src = window.URL.createObjectURL(mediaSource);
> > +{% endhighlight %}
> > +
> > +Once the MediaSource object is connected, it will perform some initialization and eventually fire a "sourceopen" event; at which point we can create a [SourceBuffer](http://www.w3.org/TR/media-source/#sourcebuffer). In the example above, we're creating an "audio/mpeg" one, which is able to parse and decode our MP3 segments; there are several [other types](http://www.w3.org/2013/12/byte-stream-format-registry/) available.
> > +
>
> Please replace quotes on sourceopen with backticks(`) to give it a
> different type face.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/WebFundamentals/pull/1381/files#r32849947>.
>

jpmedley commented on src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown:
```diff
@@ -0,0 +1,373 @@
+## Conclusion
+
+With that we've stitched all five segments seamlessly into one and have subsequently reached the end of our demo. Before we go, you may have noticed that our _onAudioLoaded()_ method has no consideration for containers or codecs... Which means all of these techniques will work irrespective of the container or codec type; below you can replay the original demo DASH-ready fragmented MP4 instead of  ...
+
```
Replace 'codecs... Which means' with 'codecs. That means'

jpmedley replied:
Also, I'm not sure that's an appropriate use of a semi-colon in the last line. Consider making two sentences.

samdutton replied:
Done x 2.

jpmedley commented on src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown:
```diff
@@ -0,0 +1,373 @@
+{% highlight bash %}
+unzip Jan_Morgenstern-Sintel-FLAC.zip
+sha1sum 1-Snow_Fight.flac
+# 0535ca207ccba70d538f7324916a3f1a3d550194  1-Snow_Fight.flac
```
The pound sign needs an escape character or this looks to markdown interpreters like a heading.

samdutton replied:
Good one – done.

jpmedley commented on src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown:
```diff
@@ -0,0 +1,373 @@
+
+## Appendix C: On Garbage Collection
+
+Memory belonging to SourceBuffers is actively [garbage collected](http://en.wikipedia.org/wiki/Garbage_collection_(computer_science)) according to content type, platform specific limits, and the current play position. In Chrome, memory will first be reclaimed from already played buffers. However, if memory usage exceeds platform specific limits, it will remove memory from unplayed buffers.
```
SourceBuffers looks to be the name of an actual object or interface. Please add back ticks.

samdutton replied:
Done.

samdutton replied:
Thanks Joe!

On 19 June 2015 at 18:46, Joe Medley <notifications@github.com> wrote:

> In
> src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown
> <https://github.com/google/WebFundamentals/pull/1381#discussion_r32851945>
> :
>
> > +
> > +    realSamples = paddedSamples - (frontPadding + endPadding);
> > +  }
> > +
> > +  return {
> > +    audioDuration: realSamples * SECONDS_PER_SAMPLE,
> > +    frontPaddingDuration: frontPadding * SECONDS_PER_SAMPLE
> > +  };
> > +}
> > +{% endhighlight %}
> > +
> > +With that we have a complete function for parsing the vast majority of gapless content out there. Edge cases certainly abound though, so caution is recommended before using similar code in production.
> > +
> > +## Appendix C: On Garbage Collection
> > +
> > +Memory belonging to SourceBuffers is actively [garbage collected](http://en.wikipedia.org/wiki/Garbage_collection_(computer_science)) according to content type, platform specific limits, and the current play position. In Chrome, memory will first be reclaimed from already played buffers. However, if memory usage exceeds platform specific limits, it will remove memory from unplayed buffers.
>
> SourceBuffers looks to be the name of an actual object or interface.
> Please add back ticks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/WebFundamentals/pull/1381/files#r32851945>.
>

petele commented:
I *believe* we can use iframes, but content must be hosted on the Google Samples site, it should not be hosted on personal sites.

petele commented on src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown:
```diff
@@ -0,0 +1,373 @@
+
+We'll get into the details of why below, but for now lets start with a demonstration. Below is the first thirty seconds of the excellent [Sintel](http://www.sintel.org/) chopped into five separate MP3 files and reassembled using MSE. The red lines indicate gaps introduced during the creation (encoding) of each MP3; you'll hear glitches at these points.
+
+<iframe seamless style="border: 0; height: 335px; width: 100%" src="https://simpl.info/mse/audio/gap"></iframe>
```
I'm OK with using iframes here, but the content needs to be served from the Web Fundamentals Samples repo, it should not be hosted on personal sites.

petele commented on src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown:
```diff
@@ -0,0 +1,373 @@
+
+Yuck! That's not a great experience; we can do better. With a little more work, using the exact same MP3 files in the above demo, we can use MSE to remove those annoying gaps. The green lines in the next demo indicate where the files have been joined and the gaps removed. On Chrome 38+ this will playback seamlessly!
+
+<iframe seamless style="border: 0; height: 335px; width: 100%" src="https://simpl.info/mse/audio/gapless"></iframe>
```
I'm OK with using iframes here, but the content needs to be served from the Web Fundamentals Samples repo, it should not be hosted on personal sites.

petele commented on src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown:
```diff
@@ -0,0 +1,373 @@
+
+First, lets backtrack and cover the basic setup of a MediaSource instance. Media Source Extensions, as the name implies, are just extensions to the existing media elements. Below, we're assigning an [Object URL](https://developer.mozilla.org/en-US/docs/Web/API/URL.createObjectURL), representing our MediaSource instance, to the source attribute of an audio element; just like you would set a standa ...
+
+{% highlight javascript %}
```
If this content is being pulled from the sample, please use the code injection stuff instead of manually including it.

samdutton replied:
Looks like the content Dale has used is somewhat different from the source code.

petele commented on src/_langs/en/updates/2015-06-12-Media-Source-Extensions-for-Audio.markdown:
```diff
@@ -0,0 +1,373 @@
+
+## Example Code
+
+{% highlight javascript %}
```
If this content is being pulled from the sample, please use the code injection stuff instead of manually including it.

petele replied:
WOW - this is a LOT of code to be showing, is it all needed? Ideally code examples should be as concise as possible showing only the absolutely necessary code. Can you reduce this to something more consumable? If developers want to consume the whole thing, they can view source on the working sample.

samdutton replied:
I think it's OK – the longer parts are mostly extended comments, and I think the examples are well constructed. 

(Note that the article is ported over from Dale's original [here](http://dalecurtis.github.io/llama-demo/index.html).)

If you feel strongly, we could ask Dale to shorten.

samdutton commented:
@gauntface @petele @PaulKinlan Not sure where to put the iframe content (nor is Matt).  

Other WF Updates have iframe content from other Google properties, e.g. HTML5 Rocks, maps.google.com, etc.

PaulKinlan commented:
_code might be ok.  Folllow the guidance of the other samples in the docs.... I think.

jpmedley commented:
Sam,

Sorry for the delay. I've been traveling. Are all the issues cleared?

petele commented:
Sam,

To include samples, create a `_code` directory in the `updates` directory and include the samples there. Then to link to them, use the following:

```
{% link_sample_button _code/sample-file.html %}
  Demo
{% endlink_sample_button %}
```

Note: `{% include_code _code/dev-orientation.html devori javascript %}` does not work right now, see issue #1384

samdutton commented:
Thanks Pete!

I've changed the iframes to includes in the .markdown file (as per your
instructions) and added src/_langs/en/updates/_code/.

In _code I've added a mse-gapless-audio directory. This has audio, css,
images and js directories for code and assets shared by the demos.

Each of the demos is in a separate directory in the demos directory.

Hope that makes sense.

As discussed, would be great if possible to include the demos directly in
iframes.

On 22 June 2015 at 10:11, Pete LePage <notifications@github.com> wrote:

> Sam,
>
> To include samples, create a _code directory in the updates directory and
> include the samples there. Then to link to them, use the following:
>
> {% link_sample_button _code/sample-file.html %}
>   Demo
> {% endlink_sample_button %}
>
> Note: {% include_code _code/dev-orientation.html devori javascript %}
> does not work right now, see issue #1384
> <https://github.com/google/WebFundamentals/issues/1384>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/WebFundamentals/pull/1381#issuecomment-114047094>
> .
>
